extensions/spring-xd-extension-kafka/src/main/java/org/springframework/integration/x/kafka/KafkaPartitionAllocator.java

```diff
@@ -0,0 +1,356 @@
+						}
+						partitions = toPartitions(partitionGroups.get(sequence - 1));
+					}
+					else {
+						if (log.isDebugEnabled()) {
+							log.debug(new String(partitionData));
+						}
+						@SuppressWarnings("unchecked")
+						List<List<Integer>> partitionGroups = objectMapper.reader(List.class).readValue(partitionData);
+						// the partition mapping is stored 0-based but sequence/count are 1-based
+						if (log.isDebugEnabled()) {
+							log.debug("Listening to: " + StringUtils.collectionToCommaDelimitedString(partitionGroups.get(sequence - 1)));
+						}
+						partitions = toPartitions(partitionGroups.get(sequence - 1));
+					}
+					return partitions;
+				}
+				finally {
+					if (partitionDataMutex.isAcquiredInThisProcess()) {
+						partitionDataMutex.release();
+					}
+				}
+			}
+			else {
+				throw new BeanInitializationException("Cannot connect to ZooKeeper, client state is " + client.getState());
+			}
+		}
+		else {
+			return partitions;
+		}
+	}
+
+	private String getDataPath() {
+		return String.format("/sources/%s/%s", streamName, moduleName);
+	}
+
+	private Partition[] toPartitions(Iterable<Integer> partitionIds) {
+		List<Partition> partitions = new ArrayList<Partition>();
+		for (Integer partitionId : partitionIds) {
+			partitions.add(new Partition(topic, partitionId));
+		}
+		return partitions.toArray(new Partition[partitions.size()]);
+	}
+
+	@Override
+	public Class<?> getObjectType() {
+		return Array.class;
+	}
+
+	@Override
+	public boolean isSingleton() {
+		return true;
+	}
+
+	@Override
+	public void onApplicationEvent(ContextClosedEvent event) {
+		CuratorFrameworkState state = client.getState();
+		if (state.equals(STARTED)) {
+			try {
+				byte[] streamStatusData;
+				try {
+					streamStatusData = client.getData().forPath(Paths.build(Paths.STREAM_DEPLOYMENTS, streamName, Paths.STATUS));
+				}
+				catch (KeeperException.NoNodeException e) {
+					// we ignore this error - the stream path does not exist, so it may have been removed already
+					// we'll behave as if we have received no data
+					streamStatusData = null;
+				}
+				DeploymentUnitStatus status = streamStatusData != null ?
+						new DeploymentUnitStatus(ZooKeeperUtils.bytesToMap(streamStatusData)) :
+						new DeploymentUnitStatus(DeploymentUnitStatus.State.undeployed);
+				if (DeploymentUnitStatus.State.undeployed.equals(status.getState())
+						|| DeploymentUnitStatus.State.undeploying.equals(status.getState())) {
+					if (client.checkExists().forPath(getDataPath()) != null) {
+						try {
+							client.delete().deletingChildrenIfNeeded().forPath(getDataPath());
+						}
+						catch (KeeperException.NoNodeException e) {
+							if (KeeperException.Code.NONODE.equals(e.code())) {
+								// ignore, most likely someone else has deleted the path already
+							}
+							else {
+								// this actually was an exception : we cannot recover, but cannot stop either
+								log.error(e);
+								e.printStackTrace();
```

Comment:
Why log the error and print a stacktrace? Won't this cause the stack trace to appear twice?